src/System.IO/src/System/IO/BinaryReader.cs

```diff
@@ -88,12 +89,15 @@ protected virtual void Dispose(bool disposing)
                     copyOfStream.Dispose();
                 }
             }
+
             _stream = null;
-            _buffer = null;
             _decoder = null;
-            _charBytes = null;
             _singleChar = null;
-            _charBuffer = null;
+            ArrayPool<byte>.Shared.Return(_buffer);
+            if (_charBytes != null)
+                ArrayPool<byte>.Shared.Return(_charBytes);
+            if (_charBuffer != null)
+                ArrayPool<char>.Shared.Return(_charBuffer);
```

Comment:
Would it make sense to return the buffer before calling dispose on the stream? This would ensure the buffers are returned even if stream.Dispose throws.